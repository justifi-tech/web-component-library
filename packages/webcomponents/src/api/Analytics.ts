import { ComponentInterface } from '@stencil/core';
import webcomponentsPackageJson from '../../package.json';
import { AnalyticsService } from './services/analytics.service';

interface iBasicData {
  resource_id: string;
  component_name: string;
  component_version: string;
  client_user_agent: string;
  client_platform: string;
  client_origin: string;
  error?: any;
}

/**
 * @class JustifiAnalytics
 * @description A class to track analytics for a component
 * @param { component } ComponentInterface - The component to track
 *
 */
class JustifiAnalytics {
  componentInstance: any;
  eventEmitters: string[];
  service: AnalyticsService;
  basicData: iBasicData;

  constructor(component: ComponentInterface) {
    this.service = new AnalyticsService();
    this.componentInstance = component;
    this.setUpBasicData();
    this.setupLifecycleTracking();
    this.trackCustomEvents();
  }

  get resourceId() {
    return (
      this.componentInstance.accountId ||
      this.componentInstance.businessId ||
      this.componentInstance.paymentId ||
      this.componentInstance.payoutId ||
      this.componentInstance.checkoutId
    );
  }

  setUpBasicData() {
    this.basicData = {
      component_name: this.componentInstance.tagName,
      component_version: webcomponentsPackageJson.version,
      client_user_agent: navigator.userAgent,
      // navigator.platform is deprecated, use navigator.userAgent instead
      client_platform: navigator.userAgent,
      client_origin: window.location.origin,
      resource_id: this.resourceId,
    };
  }

  handleCustomEvent = async (data: any) => {
    // Track the event
    await this.service.record(data);
  };

  private trackCustomEvents() {
    this.eventEmitters = ['submitted', 'error-event'];

    // for each event, add an event listener
    this.eventEmitters.forEach((eventName) => {
      // if this.componentInstance.addEventListener is a function add the event listener
      if (typeof this.componentInstance.addEventListener === 'function') {
        this.componentInstance.addEventListener(eventName, (event: any) =>
          this.handleCustomEvent({
            event_type: eventName,
            data: { ...this.basicData, error: event.detail },
          })
        );
      }
    });
  }

  setupLifecycleTracking() {
    const originalComponentDidLoad = this.componentInstance.componentDidLoad;

    this.componentInstance.componentDidLoad = () => {
      this.service.record({ event_type: 'init', data: this.basicData });

      if (originalComponentDidLoad) {
        return originalComponentDidLoad.apply(this.componentInstance);
      }
    };
  }

  cleanup() {
    // Remove event listeners
    this.eventEmitters.forEach((event) => {
      this.componentInstance.removeEventListener(event, this.handleCustomEvent);
    });
  }
}

export default JustifiAnalytics;
