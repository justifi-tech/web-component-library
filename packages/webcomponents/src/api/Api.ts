import { v4 as uuidv4 } from 'uuid';
import { PagingInfo } from './Pagination';

export interface IApiResponse<T> {
  data?: T;
  error?: IErrorObject | IServerError;
  page_info?: PagingInfo;
  errors?: string[];
  id?: number;
  type?: string;
}

export type IServerError = string;

export interface IErrorObject {
  message: string;
  code: string;
  param?: string;
}

export interface IApiResponseCollection<T> extends IApiResponse<T> {
  page_info: PagingInfo;
}

const Api = (authToken: string, apiOrigin: string) => {
  async function getAuthorizationHeader() {
    return {
      'Authorization': `Bearer ${authToken}`,
      'Idempotency-Key': uuidv4(),
      'Content-Type': 'application/json',
    };
  }

  async function makeRequest(endpoint: string, method: string, params?: any, body?: any, signal?: AbortSignal) {
    const url = `${apiOrigin}/v1/${endpoint}`;
    const requestUrl = params ? `${url}?${new URLSearchParams(params)}` : url;
    const response = await fetch(requestUrl, {
      method: method,
      headers: await getAuthorizationHeader(),
      body: body,
      signal: signal,
    });

    if (response) {
      return response.status === 204 ? {} : response.json();
    }
    handleError(requestUrl);
  }

  async function get(endpoint: string, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'GET', params, null, signal);
  }

  async function post(endpoint: string, body?: any, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'POST', params, body, signal);
  }

  async function patch(endpoint: string, body?: any, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'PATCH', params, body, signal);
  }

  async function destroy(endpoint: string, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'DELETE', params, null, signal);
  }

  return { get, post, patch, destroy };
};

function handleError(requestUrl: string): void {
  console.error(`Error fetching from ${requestUrl}`);
}

export default Api;
