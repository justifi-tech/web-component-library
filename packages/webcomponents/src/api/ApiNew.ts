import { configState } from '../components/config-provider/config-state';
import { generateId } from '../utils/utils';
import { PagingInfo } from './Pagination';

export interface IApiResponse<T> {
  data: T;
  error?: IErrorObject | IServerError;
  page_info?: PagingInfo;
  errors?: string[];
  id: number | string;
  type: string;
}

export type IServerError = string;

export interface IErrorObject {
  message: string;
  code: string;
  param?: string;
}

export interface IApiResponseCollection<T> extends IApiResponse<T> {
  page_info: PagingInfo;
}

interface RequestProps {
  endpoint: string;
  method: string;
  params?: any;
  body?: any;
  signal?: AbortSignal;
  headers?: HeadersInit;
}

const Api = (authToken: string) => {
  
  const apiOrigin = configState.apiOrigin;
  
  async function getAuthorizationHeader() {
    if (!authToken) {
      return {
        'Content-Type': 'application/json',
      };
    }

    return {
      Authorization: `Bearer ${authToken}`,
      'Idempotency-Key': generateId(),
      'Content-Type': 'application/json',
    };
  }

  async function makeRequest(props: RequestProps) {
    const { endpoint, method, params, body, signal, headers } = props;

    const url = `${apiOrigin}/v1/${endpoint}`;
    const requestUrl = params ? `${url}?${new URLSearchParams(params)}` : url;

    const defaultHeaders = await getAuthorizationHeader();
    const mergedHeaders = { ...defaultHeaders, ...headers };

    const response = await fetch(requestUrl, {
      method: method,
      headers: mergedHeaders,
      body: body,
      signal: signal,
    });

    if (response) {
      return response.status === 204 ? {} : response.json();
    }
    handleError(requestUrl);
  }

  async function get({
    endpoint,
    params,
    signal,
    headers,
  }: {
    endpoint: string;
    params?: any;
    signal?: AbortSignal;
    headers?: HeadersInit;
  }) {
    return makeRequest({ endpoint, method: 'GET', params, signal, headers });
  }

  async function post({
    endpoint,
    body,
    params,
    signal,
    headers,
  }: {
    endpoint: string;
    body?: any;
    params?: any;
    signal?: AbortSignal;
    headers?: HeadersInit;
  }) {
    return makeRequest({
      endpoint,
      method: 'POST',
      params,
      body: JSON.stringify(body),
      signal,
      headers,
    });
  }

  async function put({
    endpoint,
    body,
    params,
    signal,
  }: {
    endpoint: string;
    body?: any;
    params?: any;
    signal?: AbortSignal;
  }) {
    return makeRequest({
      endpoint,
      method: 'PUT',
      params,
      body: JSON.stringify(body),
      signal,
    });
  }

  async function patch({
    endpoint,
    body,
    params,
    signal,
  }: {
    endpoint: string;
    body?: any;
    params?: any;
    signal?: AbortSignal;
  }) {
    return makeRequest({
      endpoint,
      method: 'PATCH',
      params,
      body: JSON.stringify(body),
      signal,
    });
  }

  async function destroy({
    endpoint,
    params,
    signal,
  }: {
    endpoint: string;
    params?: any;
    signal?: AbortSignal;
  }) {
    return makeRequest({ endpoint, method: 'DELETE', params, signal });
  }

  return { get, post, put, patch, destroy };
};

function handleError(requestUrl: string): void {
  console.error(`Error fetching from ${requestUrl}`);
}

export default Api;
