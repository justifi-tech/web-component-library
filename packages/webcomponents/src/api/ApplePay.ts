import { PagingInfo } from "./Pagination";

export enum ApplePaySessionStatus {
  STATUS_SUCCESS = 'STATUS_SUCCESS',
  STATUS_FAILURE = 'STATUS_FAILURE',
}

export enum ApplePayButtonType {
  PLAIN = 'plain',
  BUY = 'buy',
  SET_UP = 'set-up',
  DONATE = 'donate',
  CHECK_OUT = 'check-out',
  BOOK = 'book',
  SUBSCRIBE = 'subscribe',
}

export enum ApplePayButtonStyle {
  BLACK = 'black',
  WHITE = 'white',
  WHITE_OUTLINE = 'white-outline',
}

export enum ApplePayMerchantCapability {
  SUPPORTS_3DS = 'supports3DS',
  SUPPORTS_EMV = 'supportsEMV',
  SUPPORTS_CREDIT = 'supportsCredit',
  SUPPORTS_DEBIT = 'supportsDebit',
}

export interface IApplePayLineItem {
  label: string;
  amount: string;
  type?: 'final' | 'pending';
}

export interface IApplePayShippingMethod {
  label: string;
  amount: string;
  identifier: string;
  detail?: string;
}

export interface IApplePayPaymentRequest {
  countryCode: string;
  currencyCode: string;
  merchantCapabilities: ApplePayMerchantCapability[];
  supportedNetworks: string[];
  total: IApplePayLineItem;
  lineItems?: IApplePayLineItem[];
  shippingMethods?: IApplePayShippingMethod[];
  applicationData?: string;
  supportedCountries?: string[];
}

export interface IApplePaySession {
  canMakePayments(): boolean;
  canMakePaymentsWithActiveCard(merchantIdentifier: string): Promise<boolean>;
  supportsVersion(version: number): boolean;
  begin(): void;
  abort(): void;
  completeMerchantValidation(merchantSession: IMerchantSession): void;
  completePayment(result: { status: ApplePaySessionStatus }): void;
  completeShippingMethodSelection(update: any): void;
  completePaymentMethodSelection(update: any): void;

  // Event handlers
  onvalidatemerchant?: (event: IApplePayValidateEvent) => void;
  onpaymentmethodselected?: (event: IApplePayMethodSelectedEvent) => void;
  onshippingmethodselected?: () => void;
  onpaymentauthorized?: (event: any) => void;
  oncancel?: (event: IApplePayCancelEvent) => void;
}

export interface IApplePayConfig {
  merchantIdentifier: string;
  displayName: string;
  initiative: string;
  initiativeContext: string;
  buttonType?: ApplePayButtonType;
  buttonStyle?: ApplePayButtonStyle;
  buttonLocale?: string;
}

export interface IApplePayError {
  code: string;
  message: string;
}

export interface IMerchantSession {
  displayName: string;
  domainName: string;
  epochTimestamp: number;
  expiresAt: number;
  merchantIdentifier: string;
  merchantSessionIdentifier: string;
  nonce: string;
  operationalAnalyticsIdentifier: string;
  pspId: string;
  retries: number;
  signature: string;
}

export interface IApplePayTokenData {
  data: string;
  header: {
    publicKeyHash: string;
    ephemeralPublicKey: string;
    transactionId: string;
  };
  signature: string;
  version: string;
}

export interface IApplePayTokenPaymentMethod {
  displayName: string;
  network: string;
  type: string;
}

export interface IApplePayToken {
  paymentData: IApplePayTokenData;
  paymentMethod: IApplePayTokenPaymentMethod;
  transactionIdentifier: string;
}

export interface IApplePayValidateEvent {
  validationURL: string;
}

export interface IApplePayCancelEvent {
  sessionError?: {
    code: 'unknown' | 'invalidMerchantSession' | 'userCancel' | string;
    info?: string;
  };
}

export interface IApplePayMethodSelectedEvent {
  paymentMethod: IApplePayTokenPaymentMethod;
}


export interface IApplePayPaymentProcessRequest {
  paymentData: IApplePayTokenData;
  paymentMethod: IApplePayTokenPaymentMethod;
  transactionIdentifier: string;
  product_details?: {
    name: string;
    price: number;
    description: string;
  };
  description?: string;
}

export interface IApplePayPaymentResponse {
  id: string;
  type: string;
  page_info: PagingInfo | null;
  data: {
    account_id: string;
    token: string;
  }
}

export interface IApplePayService {
  validateMerchant(authToken: string, accountId: string): Promise<IMerchantSession>;

  processPayment(
    authToken: string,
    accountId: string,
    payload: IApplePayPaymentProcessRequest
  ): Promise<{ success: boolean; data: IApplePayPaymentResponse }>;
}

export class ApplePayConfig implements IApplePayConfig {
  public merchantIdentifier: string;
  public displayName: string;
  public initiative: string;
  public initiativeContext: string;
  public buttonType?: ApplePayButtonType;
  public buttonStyle?: ApplePayButtonStyle;
  public buttonLocale?: string;

  constructor(data: IApplePayConfig) {
    this.merchantIdentifier = data.merchantIdentifier;
    this.displayName = data.displayName;
    this.initiative = data.initiative;
    this.initiativeContext = data.initiativeContext;
    this.buttonType = data.buttonType || ApplePayButtonType.PLAIN;
    this.buttonStyle = data.buttonStyle || ApplePayButtonStyle.BLACK;
    this.buttonLocale = data.buttonLocale || 'en-US';
  }

  public get isValid(): boolean {
    return !!(
      this.merchantIdentifier &&
      this.displayName &&
      this.initiative &&
      this.initiativeContext
    );
  }
}

export class ApplePayPaymentRequest implements IApplePayPaymentRequest {
  public countryCode: string;
  public currencyCode: string;
  public merchantCapabilities: ApplePayMerchantCapability[];
  public supportedNetworks: string[];
  public total: IApplePayLineItem;
  public lineItems?: IApplePayLineItem[];
  public shippingMethods?: IApplePayShippingMethod[];
  public applicationData?: string;
  public supportedCountries?: string[];

  constructor(data: IApplePayPaymentRequest) {
    this.countryCode = data.countryCode;
    this.currencyCode = data.currencyCode;
    this.merchantCapabilities = data.merchantCapabilities;
    this.supportedNetworks = data.supportedNetworks;
    this.total = data.total;
    this.lineItems = data.lineItems;
    this.shippingMethods = data.shippingMethods;
    this.applicationData = data.applicationData;
    this.supportedCountries = data.supportedCountries;
  }

  public get isValid(): boolean {
    return !!(
      this.countryCode &&
      this.currencyCode &&
      this.merchantCapabilities?.length &&
      this.supportedNetworks?.length &&
      this.total?.label &&
      this.total?.amount
    );
  }
}

export class ApplePayHelpers {
  static isApplePaySupported(): boolean {
    return (
      typeof window !== 'undefined' &&
      'ApplePaySession' in window &&
      window.ApplePaySession?.supportsVersion(3)
    );
  }

  static canMakePayments(): boolean {
    if (!this.isApplePaySupported()) {
      return false;
    }
    return window.ApplePaySession?.canMakePayments() || false;
  }

  static async canMakePaymentsWithActiveCard(
    merchantIdentifier: string
  ): Promise<boolean> {
    if (!this.isApplePaySupported()) {
      return false;
    }

    try {
      return (
        (await window.ApplePaySession?.canMakePaymentsWithActiveCard(
          merchantIdentifier
        )) || false
      );
    } catch (error) {
      console.error('Error checking Apple Pay active card:', error);
      return false;
    }
  }

  static formatAmount(amount: number): string {
    return (amount / 100).toFixed(2);
  }

  static parseAmount(amount: string): number {
    return Math.round(parseFloat(amount) * 100);
  }

  static createLineItem(
    label: string,
    amount: number,
    type: 'final' | 'pending' = 'final'
  ): IApplePayLineItem {
    return {
      label,
      amount: this.formatAmount(amount),
      type,
    };
  }

  static getDefaultSupportedNetworks(): string[] {
    return ['amex', 'discover', 'masterCard', 'visa'];
  }

  static getDefaultMerchantCapabilities(): ApplePayMerchantCapability[] {
    return [
      ApplePayMerchantCapability.SUPPORTS_3DS,
      ApplePayMerchantCapability.SUPPORTS_EMV,
      ApplePayMerchantCapability.SUPPORTS_CREDIT,
      ApplePayMerchantCapability.SUPPORTS_DEBIT,
    ];
  }
}

declare global {
  interface Window {
    ApplePaySession?: {
      new (
        version: number,
        paymentRequest: IApplePayPaymentRequest
      ): IApplePaySession;
      canMakePayments(): boolean;
      canMakePaymentsWithActiveCard(
        merchantIdentifier: string
      ): Promise<boolean>;
      supportsVersion(version: number): boolean;
      STATUS_SUCCESS: number;
      STATUS_FAILURE: number;
    };
  }
}
