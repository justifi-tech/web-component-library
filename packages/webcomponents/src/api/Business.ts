import { Identity, Representative } from './Identity';
import { IDocument } from './Document';
import { IBankAccount } from './BankAccount';
import { getStateAbbreviation } from '../components/business-forms/utils/helpers';

export enum BusinessClassification {
  sole_proprietor = 'sole_proprietor',
  partnership = 'partnership',
  corporation = 'corporation',
  public_company = 'public_company',
  limited = 'limited',
  non_profit = 'non_profit',
  government = 'government',
}

export interface IAddress {
  id?: string;
  platform_account_id?: string;
  line1?: string;
  line2?: string;
  postal_code?: string;
  city?: string;
  state?: string;
  country?: string;
  created_at?: string;
  updated_at?: string;
  payload?: any;
}

export class Address implements IAddress {
  public id?: string;
  public platform_account_id?: string;
  public line1?: string;
  public line2?: string;
  public postal_code?: string;
  public city?: string;
  public state?: string;
  public country?: string;
  public created_at?: string;
  public updated_at?: string;

  constructor(address: IAddress) {
    this.id = address.id;
    this.platform_account_id = address.platform_account_id;
    this.line1 = address.line1;
    this.line2 = address.line2;
    this.postal_code = address.postal_code;
    this.city = address.city;
    this.state = getStateAbbreviation(address.state);
    this.country = address.country || 'USA';
    this.created_at = address.created_at;
    this.updated_at = address.updated_at;
  }

  public get payload() {
    return {
      platform_account_id: this.platform_account_id || '',
      line1: this.line1 || '',
      line2: this.line2 || '',
      postal_code: this.postal_code || '',
      city: this.city || '',
      state: this.state || '',
      country: this.country || '',
    };
  }
}

export interface ProductCategories {
  credit: boolean;
  insurance: boolean;
  lending: boolean;
  payment: boolean;
}

export interface IAdditionalQuestions {
  business_revenue?: string;
  business_payment_volume?: string;
  business_when_service_received?: string;
  business_recurring_payments?: string;
  business_recurring_payments_percentage?: string;
  business_seasonal?: string;
  business_other_payment_details?: string;
  payload?: any;
}

export class AdditionalQuestions implements IAdditionalQuestions {
  public business_revenue?: string;
  public business_payment_volume?: string;
  public business_when_service_received?: string;
  public business_recurring_payments?: string;
  public business_recurring_payments_percentage?: string;
  public business_seasonal?: string;
  public business_other_payment_details?: string;

  constructor(additionalQuestions: IAdditionalQuestions) {
    this.business_revenue = additionalQuestions.business_revenue;
    this.business_payment_volume = additionalQuestions.business_payment_volume;
    this.business_when_service_received =
      additionalQuestions.business_when_service_received;
    this.business_recurring_payments =
      additionalQuestions.business_recurring_payments;
    this.business_recurring_payments_percentage =
      additionalQuestions.business_recurring_payments_percentage;
    this.business_seasonal = additionalQuestions.business_seasonal;
    this.business_other_payment_details =
      additionalQuestions.business_other_payment_details;
  }

  public get payload() {
    return {
      business_revenue: this.business_revenue || '',
      business_payment_volume: this.business_payment_volume || '',
      business_when_service_received: this.business_when_service_received || '',
      business_recurring_payments: this.business_recurring_payments || '',
      business_recurring_payments_percentage:
        this.business_recurring_payments_percentage || '',
      business_seasonal: this.business_seasonal || '',
      business_other_payment_details: this.business_other_payment_details || '',
    };
  }
}

export interface ICoreBusinessInfo {
  classification?: BusinessClassification;
  legal_name?: string;
  doing_business_as?: string;
  industry?: string;
  tax_id?: string;
  website_url?: string;
  email?: string;
  phone?: string;
  date_of_incorporation?: string;
}

export class CoreBusinessInfo implements ICoreBusinessInfo {
  public classification: BusinessClassification;
  public legal_name: string;
  public doing_business_as: string;
  public industry: string;
  public tax_id: string;
  public website_url: string;
  public email: string;
  public phone: string;
  public date_of_incorporation: string;

  constructor(coreBusinessInfo: ICoreBusinessInfo) {
    this.classification = coreBusinessInfo.classification;
    this.legal_name = coreBusinessInfo.legal_name;
    this.doing_business_as = coreBusinessInfo.doing_business_as;
    this.industry = coreBusinessInfo.industry;
    this.tax_id = coreBusinessInfo.tax_id;
    this.website_url = coreBusinessInfo.website_url;
    this.email = coreBusinessInfo.email;
    this.phone = coreBusinessInfo.phone;
    this.date_of_incorporation = coreBusinessInfo.date_of_incorporation;
  }

  public get payload() {
    return {
      classification: this.classification || '',
      legal_name: this.legal_name || '',
      doing_business_as: this.doing_business_as || '',
      industry: this.industry || '',
      tax_id: this.tax_id || '',
      website_url: this.website_url || '',
      email: this.email || '',
      phone: this.phone || '',
      date_of_incorporation: this.date_of_incorporation || '',
    };
  }
}

export interface IBusiness {
  additional_questions: IAdditionalQuestions;
  associated_accounts: any[];
  classification: BusinessClassification;
  bank_accounts: IBankAccount[];
  created_at: string;
  documents: IDocument[];
  doing_business_as: string;
  email: string;
  id: string;
  industry: string;
  legal_address: IAddress | {};
  legal_name: string;
  metadata: any;
  owners: Identity[];
  phone: string;
  platform_account_id: string;
  product_categories: ProductCategories;
  representative: Identity | {};
  tax_id: string;
  terms_conditions_accepted: boolean;
  updated_at: string;
  website_url: string;
  date_of_incorporation?: string;
}

export class Business implements IBusiness {
  public additional_questions: IAdditionalQuestions;
  public associated_accounts: any[];
  public classification: BusinessClassification;
  public bank_accounts: IBankAccount[];
  public created_at: string;
  public documents: IDocument[];
  public doing_business_as: string;
  public email: string;
  public id: string;
  public industry: string;
  public legal_address: Address | {};
  public legal_name: string;
  public metadata: any;
  public owners: Identity[];
  public phone: string;
  public platform_account_id: string;
  public representative: Identity | {};
  public tax_id: string;
  public terms_conditions_accepted: boolean;
  public updated_at: string;
  public website_url: string;
  public date_of_incorporation?: string;
  public product_categories: ProductCategories;

  constructor(business: IBusiness) {
    this.additional_questions = new AdditionalQuestions(
      business.additional_questions || {}
    );
    this.associated_accounts = business.associated_accounts;
    this.classification = business.classification;
    this.bank_accounts = business.bank_accounts;
    this.created_at = business.created_at;
    this.documents = business.documents;
    this.doing_business_as = business.doing_business_as;
    this.email = business.email;
    this.id = business.id;
    this.industry = business.industry;
    this.legal_address = new Address(business.legal_address || {});
    this.legal_name = business.legal_name;
    this.metadata = business.metadata;
    this.owners = business.owners;
    this.phone = business.phone;
    this.platform_account_id = business.platform_account_id;
    this.product_categories = business.product_categories;
    this.representative = new Representative(business.representative || {});
    this.tax_id = business.tax_id;
    this.terms_conditions_accepted = business.terms_conditions_accepted;
    this.updated_at = business.updated_at;
    this.website_url = business.website_url;
    this.date_of_incorporation = business.date_of_incorporation;
  }

  public get payload() {
    return {
      additional_questions: new AdditionalQuestions(this.additional_questions)
        .payload,
      classification: this.classification || '',
      doing_business_as: this.doing_business_as || '',
      email: this.email || '',
      industry: this.industry || '',
      legal_address: new Address(this.legal_address).payload,
      legal_name: this.legal_name || '',
      metadata: this.metadata || {},
      owners: this.owners.map((owner) => ({ id: owner.id })),
      phone: this.phone || '',
      platform_account_id: this.platform_account_id || '',
      representative: new Representative(this.representative).payload,
      tax_id: this.tax_id || '',
      website_url: this.website_url || '',
      date_of_incorporation: this.date_of_incorporation || '',
    };
  }
}

export interface IProductReadiness {
  business_id: string;
  created_at: string;
  id: string;
  last_verified_at: string;
  missing_optional_fields: string[];
  missing_required_fields: string[];
  percentage_complete: number;
  percentage_ready: number;
  platform_account_id: string;
  product_category: string;
  product_name: string;
  required_ready: boolean;
  updated_at: string;
}
