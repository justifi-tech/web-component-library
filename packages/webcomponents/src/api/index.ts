export * from './Api';
export * from './ComponentEvents';
export * from './ComponentError';
export {
  IBnpl,
  Completion,
  ICheckout,
  Checkout,
  ICheckoutPaymentMode,
  ICheckoutPaymentModeParam,
  ICheckoutStatus,
  CompletionStatuses,
  ICheckoutCompleteResponse,
  ILoadedEventResponse,
  CheckoutsQueryParams,
  ICheckoutPaymentMethod,
} from './Checkout';
export {
  CardBrand as CheckoutCardBrand,
  AccountType as BankAccountType,
} from './Checkout';
export * from './Insurance';
export * from './Pagination';
export * from './Payment';
export * from './PaymentBalanceTransaction';
export * from './Payout';
export * from './PayoutBalanceTransaction';
export * from './Terminal';
export * from './TerminalOrder';
export * from './TerminalModel';
export * from './SubAccount';
export * from './Document';
export * from './Refund';
export * from './Business';
export * from './Identity';
export * from './Dispute';
export * from './GrossVolume';
export * from './ApplePay';
export * from './GooglePay';
// In a future PR I would like to update the naming of the BankAccount Payment Method class in Payment.ts so this import can be updated to be written the same as the others in this file.
export { BankAccount, IBankAccount } from './BankAccount';
