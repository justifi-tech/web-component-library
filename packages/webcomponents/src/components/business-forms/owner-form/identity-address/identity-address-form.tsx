import { Component, Host, h, Prop, Watch, State } from '@stencil/core';
import { PaymentProvisioningCountryOptions } from '../../../../utils/address-form-helpers';
import { IAddress } from '../../../../api/Business';
import { getCountryFormConfig } from '../../utils';

@Component({
  tag: 'justifi-identity-address-form'
})
export class IdentityAddressForm {
  @Prop() handleFormUpdate: (values: any) => void;
  @Prop() errors: any;
  @Prop() defaultValues: any;
  @State() address: IAddress = {};
  @State() currentCountry: string = '';

  private stateControlRef!: HTMLElement;
  private postalCodeControlRef!: HTMLElement;

  componentDidLoad() {
    // Initialize currentCountry from defaultValues
    this.currentCountry = this.defaultValues?.country || '';
  }

  @Watch('address')
  handleAddressChange(newValues: any) {
    this.handleFormUpdate(newValues);
  }

  @Watch('defaultValues')
  handleDefaultValuesChange(newValues: any) {
    // Update current country when default values change
    if (newValues?.country) {
      this.currentCountry = newValues.country;
    }
  }

  private handleCountryChange = (value: string) => {
    // Clear `state` and `postal_code` fields when country changes
    this.address = {
      ...this.address,
      country: value,
      state: '',
      postal_code: '',
    };
    
    // Update state to trigger re-render for labels
    this.currentCountry = value;
    
    // Force update the form controls to show cleared values
    if (this.stateControlRef && (this.stateControlRef as any).updateInput) {
      (this.stateControlRef as any).updateInput('');
    }
    if (this.postalCodeControlRef && (this.postalCodeControlRef as any).updateInput) {
      (this.postalCodeControlRef as any).updateInput('');
    }
  }

  inputHandler = (name: string, value: string) => {
    if (name === 'country') {
      this.handleCountryChange(value);
    } else {
      // Regular field update
      this.address = {
        ...this.address,
        [name]: value
      };
    }
  }

  render() {
    const {
      regionOptions,
      regionLabel,
      postalCodeLabel,
      postalCodeConfig,
      postalCodeHelpText
    } = getCountryFormConfig(this.currentCountry || this.address?.country || this.defaultValues?.country);

    return (
      <Host>
        <div class="row gy-3">
          <div class="col-12">
            <form-control-select
              name="country"
              label="Country"
              options={PaymentProvisioningCountryOptions}
              inputHandler={this.inputHandler}
              defaultValue={this.defaultValues?.country || ''}
              errorText={this.errors?.country}
            />
          </div>
          <div class="col-12 col-md-8">
            <form-control-text
              name="line1"
              label="Street Address"
              defaultValue={this.defaultValues?.line1}
              errorText={this.errors?.line1}
              inputHandler={this.inputHandler}
              helpText="​​Enter your address exactly as stated on your driver's license. No PO Boxes."
            />
          </div>
          <div class="col-12 col-md-4">
            <form-control-text
              name="line2"
              label="Apartment, Suite, etc. (optional)"
              defaultValue={this.defaultValues?.line2}
              errorText={this.errors?.line2}
              inputHandler={this.inputHandler}
            />
          </div>
          <div class="col-12">
            <form-control-text
              name="city"
              label="City"
              defaultValue={this.defaultValues?.city}
              errorText={this.errors?.city}
              inputHandler={this.inputHandler}
            />
          </div>
          <div class="col-12 col-md-6">
            <form-control-select
              ref={(el) => this.stateControlRef = el}
              name="state"
              label={regionLabel}
              defaultValue={this.defaultValues?.state}
              options={regionOptions}
              errorText={this.errors?.state}
              inputHandler={this.inputHandler}
            />
          </div>
          <div class="col-12 col-md-6">
            <form-control-text
              ref={(el) => this.postalCodeControlRef = el}
              name="postal_code"
              label={postalCodeLabel}
              defaultValue={this.defaultValues?.postal_code}
              errorText={this.errors?.postal_code}
              maxLength={postalCodeConfig.maxLength}
              keyDownHandler={postalCodeConfig.keyDownHandler}
              inputHandler={this.inputHandler}
              helpText={postalCodeHelpText}
            />
          </div>
        </div>
      </Host>
    );
  }
}
