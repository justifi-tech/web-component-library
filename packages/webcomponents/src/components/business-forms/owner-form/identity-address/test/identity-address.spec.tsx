import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { IdentityAddressForm } from '../identity-address-form';
import { IAddress } from '../../../../../api/Business';

describe('identity-address', () => {
  it('should render IdentityAddressForm component', async () => {
    const page = await newSpecPage({
      components: [IdentityAddressForm],
      template: () => (
        <justifi-identity-address-form></justifi-identity-address-form>
      ),
    });
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();

  });

  it('should render IdentityAddressForm component with defaults', async () => {
    const businessAddress = {
      line1: 'Street 1',
      line2: 'Apartment 1',
      city: 'City',
      state: 'State',
      postal_code: '12345',
      country: 'USA',
    } as IAddress;
    const page = await newSpecPage({
      components: [IdentityAddressForm],
      template: () => (
        <justifi-identity-address-form
          defaultValues={businessAddress}
        ></justifi-identity-address-form>
      ),
    });
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();

  });
});
