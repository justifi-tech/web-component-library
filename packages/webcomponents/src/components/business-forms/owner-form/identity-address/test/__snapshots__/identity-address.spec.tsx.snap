// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identity-address should render IdentityAddressForm component 1`] = `
<justifi-identity-address-form>
  <div class="gy-3 row">
    <div class="col-12">
      <form-control-select label="Country" name="country"></form-control-select>
    </div>
    <div class="col-12 col-md-8">
      <form-control-text helptext="​​Enter your address exactly as stated on your driver's license. No PO Boxes." label="Street Address" name="line1"></form-control-text>
    </div>
    <div class="col-12 col-md-4">
      <form-control-text label="Apartment, Suite, etc. (optional)" name="line2"></form-control-text>
    </div>
    <div class="col-12">
      <form-control-text label="City" name="city"></form-control-text>
    </div>
    <div class="col-12 col-md-6">
      <form-control-select label="State" name="state"></form-control-select>
    </div>
    <div class="col-12 col-md-6">
      <form-control-text helptext="Format: 12345 or 12345-6789" label="Postal Code" maxlength="10" name="postal_code"></form-control-text>
    </div>
  </div>
</justifi-identity-address-form>
`;

exports[`identity-address should render IdentityAddressForm component with defaults 1`] = `
<justifi-identity-address-form>
  <div class="gy-3 row">
    <div class="col-12">
      <form-control-select defaultvalue="USA" label="Country" name="country"></form-control-select>
    </div>
    <div class="col-12 col-md-8">
      <form-control-text defaultvalue="Street 1" helptext="​​Enter your address exactly as stated on your driver's license. No PO Boxes." label="Street Address" name="line1"></form-control-text>
    </div>
    <div class="col-12 col-md-4">
      <form-control-text defaultvalue="Apartment 1" label="Apartment, Suite, etc. (optional)" name="line2"></form-control-text>
    </div>
    <div class="col-12">
      <form-control-text defaultvalue="City" label="City" name="city"></form-control-text>
    </div>
    <div class="col-12 col-md-6">
      <form-control-select defaultvalue="State" label="State" name="state"></form-control-select>
    </div>
    <div class="col-12 col-md-6">
      <form-control-text defaultvalue="12345" helptext="Format: 12345 or 12345-6789" label="ZIP Code" maxlength="10" name="postal_code"></form-control-text>
    </div>
  </div>
</justifi-identity-address-form>
`;
